x-pack/plugins/apm/server/lib/search_strategies/correlations/queries/query_histograms_generator.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { estypes } from '@elastic/elasticsearch';
+
+import type { ElasticsearchClient } from 'src/core/server';
+
+import type { SearchServiceFetchParams } from '../../../../../common/search_strategies/correlations/types';
+
+import type { AsyncSearchServiceLog } from '../async_search_service_log';
+import type { AsyncSearchServiceState } from '../async_search_service_state';
+import { CORRELATION_THRESHOLD, KS_TEST_THRESHOLD } from '../constants';
+
+import { getPrioritizedFieldValuePairs } from './get_prioritized_field_value_pairs';
+import { fetchTransactionDurationCorrelation } from './query_correlation';
+import { fetchTransactionDurationRanges } from './query_ranges';
+
+import type { FieldValuePairs } from './query_field_value_pairs';
+
+export async function* fetchTransactionDurationHistograms(
+  esClient: ElasticsearchClient,
+  addLogMessage: AsyncSearchServiceLog['addLogMessage'],
+  params: SearchServiceFetchParams,
+  state: AsyncSearchServiceState,
+  expectations: number[],
+  ranges: estypes.AggregationsAggregationRange[],
+  fractions: number[],
+  histogramRangeSteps: number[],
+  totalDocCount: number,
+  fieldValuePairs: FieldValuePairs
+) {
+  for (const item of getPrioritizedFieldValuePairs(fieldValuePairs)) {
+    if (
+      params === undefined ||
+      item === undefined ||
+      state.getState().isCancelled
```

Comment:
Since this is used a few times inside this for loop it might make sense to assign it to a variable outside the loop.

Reply:
The process could have been cancelled by the user in parallel while the for loop is still running, that's why we always need to check for the latest state of `isCancelled`, if we assign it once outside the loop it could be out of date. I added a helper getter so we can use `state.getIsCancelled()` in af7afce to simplify the usage a bit.